[gowthaman-01] iP

gowthaman-01 opened the pull request:
# Task Management Feature

## Description:

> "Time is what we want most, but what we use worst." ~ William Penn

This PR implements a task management feature for our application. It introduces the ability to create, manage, and track different types of tasks, including **Todo, Deadline, and Event** tasks. The feature enhances the functionality of our application by providing users with a structured way to organize their tasks 😉.

## Features:

* [x] Managing tasks
* [ ] Managing deadlines (coming soon)
* [ ] Reminders (coming soon)

## Changes Made:

* Added tasks package containing classes for Todo, Deadline, and Event tasks.
* Implemented TaskList class to manage a collection of tasks.
* Created UI components for user interaction and feedback.
* Added support for task operations such as adding, marking, unmarking, and deleting tasks.
* Implemented FileManager to handle file read/write operations for task persistence.
* Introduced Parser to parse user input and execute corresponding commands.
* Added unit tests for Task, TaskList, and Parser classes to ensure code quality and reliability.
* Refactored existing codebase for better organization and maintainability.
* Implemented DateFormatter to format task deadlines for display.
* Added MockTodo class for unit testing purposes.
* Included error handling mechanisms for invalid user input and file operations.

## How to use the application
You can view the forked repository [here](https://github.com/gowthaman-01/ip)
1. Clone the Repository:
```bash
git clone <repository-url>
```
2. Navigate to the directory where the repository is cloned.
3. Ensure you have Java installed on your machine. Compile the Java files using the following command:
```bash 
javac *.java
```
4. Run the application using the following command:
```bash 
javac Chitty
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

gowthaman-01 committed:
Rename Chatbot and add Greet, Exit functionality

gowthaman-01 committed:
Implement echo functionality

gowthaman-01 committed:
Implement ADD, LIST functionality

gowthaman-01 committed:
Implement Mark/Unmark functionality

gowthaman-01 committed:
Implement ToDos, Events and Deadlines

gowthaman-01 committed:
Remove whitespace and add hasNextLine() flag during input scan

gowthaman-01 committed:
Add Automated Text UI Testing

gowthaman-01 committed:
Handle errors

gowthaman-01 committed:
Add delete functionality

gowthaman-01 committed:
Add enums and more exceptions

gowthaman-01 committed:
Refactor code and cleanup

gowthaman-01 committed:
Implement save feature for tasks list

gowthaman-01 committed:
Implement save feature for tasks list

gowthaman-01 committed:
Merge level-7 branch to master

gowthaman-01 committed:
Implement feature to parse date and time during creation of Event and Deadline tasks

gowthaman-01 committed:
Merge level-8 branch to master

gowthaman-01 committed:
Refactor code into Command, Parser and UI classes

gowthaman-01 committed:
Refactor code into packages

gowthaman-01 committed:
Merge add-gradle-support branch to master

gowthaman-01 committed:
Automate project builds using Gradle

hiivan commented on src/main/java/helpers/DateFormatter.java:
```diff
@@ -0,0 +1,17 @@
+            return unformattedDate;
+        }
+    }
+}
```
I think you did a good job abstracting this class.

hiivan replied:
I think you did a good job abstracting the `DateFormatter` class.

hiivan commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    public Task(String description) {
+        this.description = description.trim();
+        this.done = false;
```
Try to reduce the use of `this`. Use of `this`​​ is discouraged, potentially to avoid unnecessary noise in the code. Try to avoid using this​​ except when it is shadowed by a method or constructor parameter.

hiivan commented on src/main/java/helpers/TaskList.java:
```diff
@@ -0,0 +1,131 @@
+        for (String taskString: taskStrings) {
+            String[] taskDetails = taskString.split("\\|");
+            Task newTask;
+            switch (taskDetails[0]) {
```
Is this the right indentation for `switch` and `case`?

hiivan reviewed (commented):
Generally good OOP principles. Just minor coding practices that need to be followed. If not, good job!

hiivan commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public abstract class Task {
+    protected final String description;
+    private boolean done;
```
Is this the best way to name a boolean variable? Perhaps something more explicit like `isDone`?

hiivan commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,53 @@
+    public boolean getIsDone() {
+        return this.done;
+    }
+
```
Try to reduce the use of `this`. Use of `this` is discouraged, potentially to avoid unnecessary noise in the code. Try to avoid using `this` except when it is shadowed by a method or constructor parameter.

hiivan commented on src/main/java/Chitty.java:
```diff
@@ -0,0 +1,44 @@
+                ui.displayInvalidCommandMessage();
+            } catch (Exception exception) {
+                ui.displayInvalidInputMessage();
+            }
```
I like your approach to error handling here. Very robust.

hiivan commented on src/main/java/helpers/Parser.java:
```diff
@@ -0,0 +1,35 @@
+            case DELETE:
+                return new DeleteCommand(Integer.parseInt(input[1]));
+            default:
+                return new ExitCommand();
```
Is this the right indentation for `switch` and `case`?

hiivan reviewed (commented):
Generally good OOP principles. Just minor coding practices that need to be followed. If not, good job!

gowthaman-01 committed:
Implement unit tests

gowthaman-01 committed:
Generate JAR file

gowthaman-01 committed:
Add documentation to classes and methods

gowthaman-01 committed:
Merge A-JavaDoc branch to master

gowthaman-01 committed:
Refactor code to match coding standards

gowthaman-01 committed:
Merge A-CodingStandard branch to master

gowthaman-01 committed:
Rename packages to follow best naming practices

gowthaman-01 committed:
Address PR comments

gowthaman-01 committed:
Modify J-Docs to improve clarity

gowthaman-01 committed:
Implement search feature

gowthaman-01 committed:
Merge Level-9 branch to master

gowthaman-01 committed:
Use checkStyle to detect coding style violations

gowthaman-01 committed:
Merge branch A-CheckStyle into master

Morninggloryyy commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,114 @@
+     * @param ui                            The user interface object.
+     * @throws InvalidInputFormatException  If the input format is invalid.
+     */
+    private void addDeadline(String[] deadlineDetails, TaskList taskList, Ui ui) throws InvalidInputFormatException {
```
I think it might be better to try to keep line length shorter than 110 characters. :))

Morninggloryyy commented on src/main/java/utils/Parser.java:
```diff
@@ -0,0 +1,48 @@
+    public static Command parse(String[] input) throws IllegalArgumentException {
+        String commandString = input[0];
+        CommandType command = CommandType.valueOf(commandString.toUpperCase());
+        switch (command) {
+        case LIST:
+            return new ListCommand();
+        case TODO:
+            return new AddCommand(CommandType.TODO, input);
+        case DEADLINE:
+            return new AddCommand(CommandType.DEADLINE, input);
+        case EVENT:
+            return new AddCommand(CommandType.EVENT, input);
+        case MARK:
+            return new MarkCommand(Integer.parseInt(input[1]));
+        case UNMARK:
+            return new UnmarkCommand(Integer.parseInt(input[1]));
+        case DELETE:
+            return new DeleteCommand(Integer.parseInt(input[1]));
+        case FIND:
+            return new FindCommand(input[1]);
+        default:
+            return new ExitCommand();
+        }
```
As shown in Java coding standard, "The explicit //Fallthrough comment should be included whenever there is a case statement without a break statement.". So, it might be better to make some changes regarding it? :))

JonChong98 reviewed (commented):
Couldn't find any glaring issues for coding standard aside from the few minor ones already highlighted by fellow reviewer. Good job!

gowthaman-01 committed:
Add GUI to the chatbot

gowthaman-01 committed:
Merge branch Level-10 into master

gowthaman-01 committed:
Remove I/O redirection automated testing

gowthaman-01 committed:
Adding assertions.

The current situation involves the initialization of resources in the Main class without any checks for null values.

It needs to change because without proper checks, there's a risk of encountering NullPointerExceptions during application execution, especially if essential resources are not loaded correctly.

To address this issue, I have added assertion checks in the Main class to ensure that FXMLLoader location and image resources are not null before proceeding with application setup.

This change is made to enhance the reliability and robustness of the application initialization process. By ensuring that essential resources are properly loaded, we can mitigate potential runtime errors and improve the overall stability of the application.

gowthaman-01 committed:
Merge branch A-Assertions into master

gowthaman-01 committed:
Improve code quality in Main class

The current code in the Main class could benefit from enhancements to improve its overall quality.

* Improving code quality helps in reducing bugs, improving performance, and facilitating future development and maintenance.
* Consistent use of coding standards and conventions contributes to better collaboration among developers and ensures a more uniform codebase.
* For further discussion or suggestions on improving code quality, feel free to reach out or refer to relevant resources on code quality practices.

gowthaman-01 committed:
Merge pull request #2 from gowthaman-01/A-CodeQuality

Improve code quality in Main class

gowthaman-01 committed:
Update build.gradle

gowthaman-01 committed:
Set up CI

gowthaman-01 committed:
Implement C-Help extension

gowthaman-01 committed:
Implement better GUI and chatbot personality

gowthaman-01 committed:
Update GUI and upload Ui.png screenshot

gowthaman-01 committed:
Update task_list.txt

gowthaman-01 committed:
Update README.md

gowthaman-01 committed:
Update README.md
